contrib/src/compression.rs

```diff
@@ -0,0 +1,169 @@
+//! This module provides brotli ang gzip compression for all non-image
+//! responses without Content-Encoding header or with Content-Encoding br or
+//! gzip respectively for requests that send Accept-Encoding br and gzip. If
+//! accepted, brotli compression is preferred over gzip.
+//!
+//! To add this feature to your Rocket application, use
+//! .attach(rocket_contrib::Compression::fairing())
+//! to your Rocket instance. Note that you must add the
+//! "compression" feature for brotli and gzip compression to your rocket_contrib
+//! dependency in Cargo.toml. Additionally, you can load only brotli compression
+//! using "brotli_compression" feature or load only gzip compression using
+//! "gzip_compression" in your rocket_contrib dependency in Cargo.toml.
+//!
+//! In the brotli algorithm, quality is set to 2 in order to have really fast
+//! compressions with compression ratio similar to gzip. Also, text and font
+//! compression mode is set regarding the Content-Type of the response.
+//!
+//! In the gzip algorithm, quality is the default (9) in order to have good
+//! compression ratio.
+//!
+//! For brotli compression, the rust-brotli crate is used.
+//! For gzip compression, flate2 crate is used.
+
+use std::io::Cursor;
+#[cfg(feature = "gzip_compression")]
+use std::io::Write;
+
+use rocket::{Request, Response};
+use rocket::http::{Header, HeaderMap};
+use rocket::fairing::{Fairing, Info, Kind};
+
+#[cfg(feature = "brotli_compression")]
+use brotli;
+#[cfg(feature = "brotli_compression")]
+use brotli::enc::backward_references::BrotliEncoderMode;
+
+#[cfg(feature = "gzip_compression")]
+use flate2;
+#[cfg(feature = "gzip_compression")]
+use flate2::write::GzEncoder;
+
+#[derive(Debug, Default)]
+pub struct Compression;
+
+impl Compression {
+    /// This function creates a Compression to be used in your Rocket
+    /// instance. Add ```.attach(rocket_contrib::Compression::fairing())```
+    /// to your Rocket instance to use this fairing.
+    ///
+    /// # Returns
+    ///
+    /// A Compression instance.
+    pub fn fairing() -> Compression {
+        Compression::default()
+    }
+}
+
+impl Fairing for Compression {
+    fn info(&self) -> Info {
+        Info {
+            name: "Brotli and gzip compressors for responses",
+            kind: Kind::Response,
+        }
+    }
+
+    fn on_response(&self, request: &Request, response: &mut Response) {
+        let accept_headers: Vec<&str> = request.headers().get("Accept-Encoding").collect();
```

Comment:
This line and the series of booleans below are unnecessary. 

The first thing this should do is check whether this is an image. The `ContentType` header is cached in the request, so you can do this really easy and correctly with:

```rust
// We don't compress images, so bail early.
if request.content_type().top() == "image" {
    return;
}
```

Then we simply want to check whether `br` or `gzip` are accepted content-encoding and perform GZIP or BR compression if so, preferring to perform brotli compression if it's accepted. This is can be done simpler as well:

```rust
let accepted_encodings = request.headers()
    .get("Accept-Encoding")
    .flat_map(|e| e.split(","));

#[cfg(feature = "brotli_compression")]
if accepted_encodings.any(|e| e == "br") {
    // Do brotli compression and return.
}

#[cfg(feature = "brotli_compression")]
if accepted_encodings.any(|e| e == "gzip") {
    // Do GZIP compression and return.
}
```